src/zodbpickle/_pickle_27.c

```diff
@@ -3929,19 +3940,30 @@ load_short_binbytes(Unpicklerobject *self)
 
     if (self->read_func(self, &s, l) < 0) return -1;
 
-    if (!( py_string = PyString_FromStringAndSize(s, l)))  return -1;
+    if (!( py_string = PyString_FromStringAndSize(s, l)))
+        return -1;
 
     if (!( args = PyTuple_New(1) ))
-        return -1;
+        goto Cleanup;
 
     if (!(PyTuple_SET_ITEM(args, 0, py_string)))
-        return -1;
+		/* practically speaking this macro cannot fail */
+        goto Cleanup;
 
     if (!( py_binary = PyObject_CallObject(BinaryType, args)))
-        return -1;
+        goto Cleanup;
 
-    PDATA_PUSH(self->stack, py_binary, -1);
-    return 0;
+Cleanup:
+	if (!args) {
+		/* The tuple steals the string reference */
+		Py_XDECREF(py_string);
+	}
+    Py_XDECREF(args);
+    if ( py_binary ) {
+        PDATA_PUSH(self->stack, py_binary, -1);
+        return 0;
+    }
+    return -1;
 }
 
```

Comment:
Hi again ;) 
Thanks a lot for the leak deep tracking !

I would rewrite it this way from my point of view. 
`````

    if (!( py_string = PyString_FromStringAndSize(s, l)))
        return -1;

    if (!( args = PyTuple_New(1) ))
        goto err;
 
    if (!(PyTuple_SET_ITEM(args, 0, py_string)))
        goto err;

    if (!( py_binary = PyObject_CallObject(BinaryType, args)))
        goto err;
 
   /* Clean Up PyTuple_New */
    Py_DECREF(args);
    PDATA_PUSH(self->stack, py_binary, -1);
    return 0
 
err:
    /* In case PyTuple_New fail */
    Py_XDECREF(py_string);
    Py_XDECREF(args);
    return -1;
```

Reply:
That code will crash if `PyObject_CallObject` fails because `py_string` will be deallocated twice---as you pointed out, `args` steals the reference. Eventually you'll get a crash like this:
```
0   org.python.python             	0x0000000104ac6edc PyObject_Malloc + 76
1   org.python.python             	0x0000000104acc8be PyString_FromStringAndSize + 174
2   org.python.python             	0x0000000104acdaaa PyString_DecodeEscape + 106
3   _pickle.so                    	0x0000000106389e74 load_string + 260
```

I prefer the explicit test for the error condition as written.